### ISSUE TITLE
RawHash on R10.4 human genome

### ISSUE THREAD

**Issue by hasindu2008** [CMU-SAFARI/RawHash#5]
Hello

I wanted to try out RawHash on some human genome data. I am still at the first step on generating the index.
For R9.4 I could create the index using the following command without any issue.
```
bin/rawhash2 -d ref.ind -p extern/kmer_models/legacy/legacy_r9.4_180mv_450bps_6mer/template_median68pA.model -t 32 /genome/hg38noAlt.fa
```

however, for R10.4, I am getting an error.
```
bin/rawhash2 -d ref.ind -p extern/kmer_models/dna_r10.4.1_e8.2_400bps/9mer_levels_v1.txt -t 32 /genome/hg38noAlt.fa    Error: cannot convert '
' to float
free(): invalid next size (normal)
Aborted (core dumped)
```

Any idea what I am doing wrong?

**Comment by hasindu2008** [comment]
Until I  get rawHash working on human genome for R10 data, I thought of starting with a small genome with R9 data first. 
I am basically trying to evaluate the signal simulator  [squigulator](https://github.com/hasindu2008/squigulator) with RawHash mapping (and vice versa).

```
squigulator test/nCoV-2019.reference.fasta -o nCoV-2019.blow5 -n 200000 -t32  -c out.paf --paf-ref
 rawhash2 -p extern/kmer_models/legacy/legacy_r9.4_180mv_450bps_6mer/template_median68pA.model -t 32 -x fast nCoV-2019.reference.fasta nCoV-2019.blow5  > a.paf
```

I am observing that rawhash2 is using very little of the CPU despite providing 32 threads. Is that expected or is it because every part being not multi-threaded?

**Comment by canfirtina** [comment]
Hi @hasindu2008,

For your first message:

You need to specify `--r10` when using R10 datasets:

``bin/rawhash2 --r10 -d ref.ind -p extern/kmer_models/dna_r10.4.1_e8.2_400bps/9mer_levels_v1.txt -t 32 /genome/hg38noAlt.fa``

Please note that RawHash2 should provide relatively good accuracy for R10.4 while for R10.4.1, the accuracy is not good currently. Please see the corresponding discussion in our recent preprint about this (https://arxiv.org/pdf/2309.05771v4 Section 3.5).

I will now update the main README to make it more clear when using R10 datasets. Thanks for the feedback.

For your second message:

This is expected because of the pipelining strategy in RawHash2. Similar to minimap2, several steps are pipelined: 1) partially reading from file (as determined by `-K`), 2) mapping, 3) writing. Although the IO step (either reading or writing) can be processed independently using one additional thread (if multiple threads are available with `-t`) to overlap the IO latency with mapping, IO part can still take much longer time than mapping. This means that CPU uses mostly single thread in its overall execution time due to being bottlenecked by IO. This usually happens if the mapping step is very cheap (e.g., you are working with bacterial genomes).

To avoid this, I would suggest increasing the minibatch size (`-K`). This can increase the peak memory usage but then it will enable utilizing multiple threads better.

**Comment by hasindu2008** [comment]
Hi @canfirtina 

Thanks for the prompt answer. I will stick to R9.4.1 for the moment then.

For the second question, I just tried with a huge batch size of 10G. However, the CPU utilisation is only only 161%, which should have ideally been close to 3200%.
```
        Command being timed: "bin/rawhash2 -p extern/kmer_models/legacy/legacy_r9.4_180mv_450bps_6mer/template_median68pA.model -t 32 -x fast /home/hasindu/hasindu2008.git             /squigulator/test/nCoV-2019.reference.fasta /home/hasindu/hasindu2008.git/squigulator/nCoV-2019.blow5 -K10G"
        User time (seconds): 322.80
        System time (seconds): 40.66
        Percent of CPU this job got: 161%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 3:44.48
```

I do not think that this is a I/O bottleneck as I am not using a FAST5, instead a BLOW5.  The same BLOW5 when I run through my benchmark programme that reads the slow5 and converts it to pico-ampere (which I believe is the same you would be doing in rawhash) using the same number of threads (32):
```
Time for disc reading 5.289125
Time for getting samples (disc+depress+parse) 17.905813
```
```
        User time (seconds): 518.51
        System time (seconds): 26.69
        Percent of CPU this job got: 2459%
        Elapsed (wall clock) time (h:mm:ss or m:ss): 0:22.16
```
Only 5.3 s has been spent on actual disk I/O. 
When including disk I/O with the time spent on decompression and parsing this becomes 17.8 seconds. For the whole programme that does all above + convert to pA it takes 22 seconds with 2459% CPU usage.


Also when reading through the manuscript you shared, I saw the this table and I find it a bit odd.
![image](https://github.com/CMU-SAFARI/RawHash/assets/12987163/6520bde4-98a7-45bb-a0ae-83682d6e0f5a), especially that the performance of BLOW5 is quite slower than it should be for a case like this. 
Could you link me to the two datasets you used here for E. coli and Yeast?

**Comment by hasindu2008** [comment]
So, I ran the same above slow5 benchmark with a single thread and it takes 3 min 48.81 sec, which is close to the RawHash, which made me think that you are still probably using BLOW5 in a single-threaded fashion. Going through the code I can see it is indeed single-threaded https://github.com/CMU-SAFARI/RawHash/blob/ad5bf88e34b5c6d1f5c33bbae0ba2c6bd48ce534/src/rsig.c#L488. 
But seems like for POD5, you are reading them as batches https://github.com/CMU-SAFARI/RawHash/blob/ad5bf88e34b5c6d1f5c33bbae0ba2c6bd48ce534/src/rsig.c#L460 which could be multi-threaded internally in the POD5 library. 


I suggest to at least use this [slow5_get_next_batch function](https://hasindu2008.github.io/slow5lib/slow5_api/mt_api/slow5_get_next_batch.html) in slow5.  This method uses easy multi-threaded API in SLOW5. The one I that would be even more efficient is the use of low-level slow5 functions such as the use of `slow5_get_next_bytes` for reading raw bytes from the disk at the I/O stage and then doing the decompression and parsing at the processing step (for example just before doing event detection) using slow5_decode with multiple threads. A pthread example is [here](https://github.com/hasindu2008/slow5lib/blob/master/examples/adv/sequential_read_pthreads.c) and an openMP example is [here](https://github.com/hasindu2008/slow5lib/blob/master/examples/adv/sequential_read_openmp.c).  


The other thing you should consider is the type of compression used in file formats. FAST5 supports zlib and vbz (zstd+streamvbyte) as compression methods. Default used to be zlib which ONT at once point made vbz  the default.
BLOW5 also has zlib and zstd+streamvbyte and the default is zlib. POD5 only has vbz  (actually an SIMD accelerated version of vbz). zlib is much more CPU-intensive than zstd.  Unless you match the compression type in these formats, the comparison is not going to be sound. For instance, when I run my slow5 I/O benchmark using a single  thread, given below is the time for disk reading + decompression + parsing for zlib vs vbz (zstd).  You can use the `-c zstd` option in slow5tools during conversion (or can use view to convert a zlib BLOW5 to zstd BLOW5)  to generate a zstd based BLOW5. 


```
zlib:
Time for disc reading 4.579812
Time for getting samples (disc+depress+parse) 142.601073

zstd:
Time for disc reading 4.750786
Time for getting samples (disc+depress+parse) 55.587349
```


You can add `slow5_mt=1 and zstd=1`  to enable multithreading as well as zstd in slow5lib.
https://github.com/CMU-SAFARI/RawHash/blob/ad5bf88e34b5c6d1f5c33bbae0ba2c6bd48ce534/src/Makefile#L185

**Comment by canfirtina** [comment]
Thanks for the response @hasindu2008. I have applied your suggestions. Multi-threading and reading in batches:

https://github.com/CMU-SAFARI/RawHash/blob/01f31fe0e82a638940fc08b4e4c49ddabc247686/src/rsig.c#L192-L194

I have also enabled zstd from Makefile as you suggested.

A couple of questions:

**First question:** Should we be explicitly keep getting the next batch from the same file in a while loop like this?
https://github.com/hasindu2008/slow5lib/blob/e5fbd7558fbbb76fce2d6ea127c0e6e8f30cd3ce/examples/mt/mt.c#L43

As opposed to using a while loop, we currently read a single batch:
https://github.com/CMU-SAFARI/RawHash/blob/01f31fe0e82a638940fc08b4e4c49ddabc247686/src/rsig.c#L194

And then process the records from that batch:
https://github.com/CMU-SAFARI/RawHash/blob/01f31fe0e82a638940fc08b4e4c49ddabc247686/src/rsig.c#L490-L503

I tried the loop idea but it kept crashing. I was also confused why ret should not be smaller than the batch size:
https://github.com/hasindu2008/slow5lib/blob/e5fbd7558fbbb76fce2d6ea127c0e6e8f30cd3ce/examples/mt/mt.c#L51

Because from the inner loop I understand ret is the number of reads fetched from the latest batch and can it not be smaller than the batch size. Maybe I am missing something.
https://github.com/hasindu2008/slow5lib/blob/e5fbd7558fbbb76fce2d6ea127c0e6e8f30cd3ce/examples/mt/mt.c#L45

**Second question:** I went over the issue you created here: https://github.com/nanoporetech/pod5-file-format/issues/1
I have looked into this myself as well but could not find a trace to confirm that POD5 (or Arrow) is aggressively using more than one thread, unless we specifically design it to do so (as you discussed in the issue you created above). Do you have more insights about this?

**Comment by hasindu2008** [comment]
Hi @canfirtina 

First question: 

So in your current implementation do you call `slow5_get_next_batch` only once per file? If that is the case, if a file has more than 4096 reads, only the first 4096 will be taken. `slow5_get_next_batch` has to be repeatedly called until the return value is less than the requested batch size. That https://github.com/hasindu2008/slow5lib/blob/e5fbd7558fbbb76fce2d6ea127c0e6e8f30cd3ce/examples/mt/mt.c#L51 comment is indeed confusing. My apologies, I should polish up that comment. As stated under return value in https://hasindu2008.github.io/slow5lib/slow5_api/mt_api/slow5_get_next_batch.html, the slow5_get_next_batch() will return the number of records it fetched. This should equal the requested batch `num_reads` except for the last batch in the file. Suppose the file had 10,000 reads and we request 4000 reads using slow5_get_next_batch repeatedly. The return values will be 4000,4000,2000. The fact that the batch size was smaller than what was requested (2000 in this case), can be used to know that the file has reached the end. Did I clarify your question or did I misunderstand your question? 

Do you have the loop method that crashes, in a separate branch or so? I can have a look.


Second question: 
This issue was in 2022 when I tried to do some preliminary benchmarks with POD5. But since then, there have been so many changes to the POD5 structure and the libraries. 

At that time, I observed that the CPU usage went >100% when running supposedly at a single thread.  So while the POD5 wrapper might be single-threaded, the underlying arrow library is possibly multi-threaded. As per this https://arrow.apache.org/docs/cpp/threading.html _"The size of the I/O thread pool currently defaults to 8 threads"_ and _"However, many operations will continue to use the I/O thread pool. This means that some parallelism may still occur even when serial execution is requested."._  So it seems like there is some underlying threading in the arrow.

However, I did not look into the code given the complex nature of the arrow library and the POD5 wrapper.
Also, not sure if (3)  in https://github.com/nanoporetech/pod5-file-format/issues/1#issuecomment-1134485782 is implemented or not.

**Comment by canfirtina** [comment]
I am closing this issue as we hopefully found the answers to both questions during our in person discussion :-)

Feel free to re-open it if you want to continue the discussion or ask questions related to this @hasindu2008.

Summary: The latest commit resolves the issue we had when processing batches in a loop. The issue was we were trying to run `slow5_get_next_batch` even though there was no next batch. If the returned length is less than the batch size, it is for now safe to assume that this is the last batch.

**Comment by hasindu2008** [comment]
Thanks

**Comment by daanishmahajan** [comment]
Hi @canfirtina, I tried running the tool using --r10 command but I am getting the following message corresponding to every read in the err log file, because of which the output PAF file is coming out to be empty (no alignment information is displayed). Where am I going wrong?

**Command:**
rawhash2 --r10 -t 32 -p $KMER_MODEL -o $TMAP_OUT -x sensitive --max-chunks $i -d $REF_OUT $TDIR
    
**Message:**
[WARNING][1;31m failed to parse the FASTA/FASTQ record next to 'sequencing_kit'. Continue anyway.[0m
[WARNING][1;31m failed to parse the first FASTA/FASTQ record. Continue anyway.[0m
[WARNING][1;31m empty sequence name in the input.[0m
[WARNING][1;31m failed to parse the FASTA/FASTQ record next to ''. Continue anyway.[0m
[WARNING][1;31m failed to parse the first FASTA/FASTQ record. Continue anyway.[0m
[WARNING][1;31m failed to parse the FASTA/FASTQ record next to '"�;�@�O��y��'L�[���_���'��h4WɦY��)~|���3��_�'. Continue anyway.[0m
[WARNING][1;31m failed to parse the FASTA/FASTQ record next to 'S���l��7�]�r@����[dEDl@�w�a`��jA��L���Ϭߜy�~`�JXE�˱��e�S����x��vҟ�s-4�ߐY���M�����lp>�'. Continue anyway.[0m
[WARNING][1;31m failed to parse the first FASTA/FASTQ record. Continue anyway.[0m
[M::main::0.326*0.45] loaded/built the index for 13 target sequence(s)
[M::ri_idx_stat] pore kmer size: 9; concatanated events: 8; quantization bits: 4; w: 0; n: 0; #seq: 13
[INFO] No files to query index on. Only the index is constructed.

**Comment by canfirtina** [comment]
Hi @daanishmahajan,

The following script shows the commands we use for running RawHash2: https://github.com/CMU-SAFARI/RawHash/blob/main/test/scripts/run_rawhash2.sh.

Based on the command you provided, I believe you are trying to run the index-only mode (with -d) by writing the index to $REF_OUT. In this mode, RawHash2 expects the last argument to be a FASTA/FASTQ file. It is likely that you are providing a directory, which is not the expected input. Something like this could work:

`rawhash2 --r10 -t 32 -p $KMER_MODEL -x sensitive --max-chunks $i -d $REF_OUT <FASTA File>`

After completing the above indexing step, you can use the index you created ($REF_OUT) in mapping by providing it as input (assuming $TDIR is a directory containing signal files):

`rawhash2 --r10 -t 32 -p $KMER_MODEL -x sensitive --max-chunks $i $REF_OUT $TDIR`

Please feel free to create a new issue if the problem persists.

**Comment by daanishmahajan** [comment]
Hi @canfirtina,
It worked! Thanks a lot :). The mistake I was doing is I was not passing --r10 while creating the index and only while mapping.
